Firestore/core/src/firebase/firestore/local/local_store.mm

```diff
@@ -0,0 +1,486 @@
+    // the remote documents in advance in a single call.
+    OptionalMaybeDocumentMap existing_docs =
+        remote_document_cache_->GetAll(updated_keys);
+
+    for (const auto& kv : remote_event.document_updates()) {
+      const DocumentKey& key = kv.first;
+      const MaybeDocument& doc = kv.second;
+      absl::optional<MaybeDocument> existing_doc;
+      auto found_existing = existing_docs.get(key);
+      if (found_existing) {
+        existing_doc = *found_existing;
+      }
+
+      // Note: The order of the steps below is important, since we want to
+      // ensure that rejected limbo resolutions (which fabricate NoDocuments
+      // with SnapshotVersion::None) never add documents to cache.
+      if (doc.type() == MaybeDocument::Type::NoDocument &&
+          doc.version() == SnapshotVersion::None()) {
+        // NoDocuments with SnapshotVersion::None are used in manufactured
+        // events. We remove these documents from cache since we lost access.
+        remote_document_cache_->Remove(key);
+        changed_docs = changed_docs.insert(key, doc);
+      } else if (!existing_doc || doc.version() > existing_doc->version() ||
+                 (doc.version() == existing_doc->version() &&
+                  existing_doc->has_pending_writes())) {
+        // TODO(index-free): Comment in this assert when we enable Index-Free
+        // queries HARD_ASSERT(remoteEvent.snapshot_version() !=
+        // SnapshotVersion::None(),
+        //            "Cannot add a document when the remote version is zero");
+        remote_document_cache_->Add(doc);
+        changed_docs = changed_docs.insert(key, doc);
+      } else {
+        LOG_DEBUG("FSTLocalStore Ignoring outdated watch update for %s. "
+                  "Current version: %s  Watch version: %s",
+                  key.ToString(), existing_doc->version().ToString(),
+                  doc.version().ToString());
+      }
+
+      // If this was a limbo resolution, make sure we mark when it was accessed.
+      if (limbo_documents.contains(key)) {
+        persistence_->reference_delegate()->UpdateLimboDocument(key);
+      }
+    }
+
+    // HACK: The only reason we allow omitting snapshot version is so we can
+    // synthesize remote events when we get permission denied errors while
+    // trying to resolve the state of a locally cached document that is in
+    // limbo.
+    const SnapshotVersion& remote_version = remote_event.snapshot_version();
+    if (remote_version != SnapshotVersion::None()) {
+      HARD_ASSERT(remote_version >= last_remote_version,
+                  "Watch stream reverted to previous snapshot?? (%s < %s)",
+                  remote_version.ToString(), last_remote_version.ToString());
+      query_cache_->SetLastRemoteSnapshotVersion(remote_version);
+    }
+
+    return local_documents_->GetLocalViewOfDocuments(changed_docs);
+  });
+}
+
+bool LocalStore::ShouldPersistQueryData(const QueryData& new_query_data,
```

Comment:
This function looks like it should be `const`.

Reply:
Done.